### PR TITLE
feat: migrate serverless instance to the new SDK

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ [INTMDB-1106](https://jira.mongodb.org/browse/INTMDB-1106)

Please include a summary of the fix/feature/change, including any relevant motivation and context.

Link to any related issue(s): 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Manual QA performed:

- [x] cfn invoke for each of CRUDL/cfn test
- [x] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [x] Published to AWS private registry
- [x] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [x] Deleted stack to ensure resources are deleted
- [x] Created multiple resources in same stack
- [x] Validated in Atlas UI
- [x] Included screenshots

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code
- [x] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

### CFN test

```bash
cfn test --function-name TestEntrypoint --verbose

== 10 passed, 2 skipped, 12 deselected in 1019.08s (0:16:59) ==
```
### CFN Stack
Screenshot 2023-09-29 at 15 22 07

Screenshot 2023-09-29 at 15 23 09

### CODE REVIEW

File: cfn-resources/project-ip-access-list/cmd/resource/create.go [hunk 1]
```diff
@@ -18,9 +18,7 @@ import (
 	"context"
 
 	"github.com/aws-cloudformation/cloudformation-cli-go-plugin/cfn/handler"
-	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/cloudformation"
-	"github.com/mongodb/mongodbatlas-cloudformation-resources/profile"
 	"github.com/mongodb/mongodbatlas-cloudformation-resources/util"
 	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/logger"
 	progressevents "github.com/mongodb/mongodbatlas-cloudformation-resources/util/progressevent"
```
Comment (lantoli) [inline cfn-resources/project-ip-access-list/cmd/resource/create.go:24]: removing alias and use progressevent

File: cfn-resources/serverless-instance/cmd/resource/resource.go [hunk 2]
```diff
@@ -74,28 +70,30 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
-		ProviderSettings:             setProviderSettings(currentModel),
+		ProviderSettings:             *setProviderSettings(currentModel),
 		ServerlessBackupOptions:      setBackupOptions(currentModel),
 		TerminationProtectionEnabled: currentModel.TerminationProtectionEnabled,
 	}
 
-	serverless, res, err := client.ServerlessInstances.Create(context.Background(), *currentModel.ProjectID, serverlessInstanceRequest)
+	serverless, res, err := client.AtlasV2.ServerlessInstancesApi.CreateServerlessInstance(context.Background(), *currentModel.ProjectID, serverlessInstanceRequest).Execute()
 	if err != nil {
 		_, _ = log.Warnf("Serverless - Create() - error: %+v", err)
```
Comment (lantoli) [inline cfn-resources/serverless-instance/cmd/resource/resource.go:82]: is it needed? if so i think it's better debugf

File: cfn-resources/project-ip-access-list/cmd/resource/create.go (hunk 1 above)
Comment (andreaangiolillo) [reply cfn-resources/project-ip-access-list/cmd/resource/create.go:25]: Updated

File: cfn-resources/serverless-instance/cmd/resource/resource.go [hunk 3]
```diff
@@ -307,59 +310,58 @@ func readServerlessInstance(cluster *mongodbatlas.Cluster, profile *string) (ser
...
 	if cluster.ConnectionStrings != nil {
 		serverless.ConnectionStrings = &ServerlessInstanceConnectionStrings{
-			StandardSrv:     &cluster.ConnectionStrings.StandardSrv,
+			StandardSrv:     cluster.ConnectionStrings.StandardSrv,
 			PrivateEndpoint: readPrivateEndpoint(cluster.ConnectionStrings.PrivateEndpoint),
 		}
 	}
-	serverless.CreateDate = &cluster.CreateDate
-	serverless.MongoDBVersion = &cluster.MongoDBVersion
+	serverless.CreateDate = admin.PtrString(cluster.CreateDate.String())
```
Comment (lantoli) [inline cfn-resources/serverless-instance/cmd/resource/resource.go:313]: what is PtrString used for? what are serverless.CreateDate and cluster.CreateDate types?

Comment (andreaangiolillo) [reply cfn-resources/serverless-instance/cmd/resource/resource.go:313]: it returns a `*string` given the returned value from `cluster.CreateDate.String() `

Comment (lantoli) [reply cfn-resources/serverless-instance/cmd/resource/resource.go:313]: what about this so you also make sure the date format is correct?   serverless.CreateDate = util.TimePtrToStringPtr(cluster.CreateDate)

Comment (lantoli) [reply cfn-resources/serverless-instance/cmd/resource/resource.go:313]: we don't want to use String() function in time.Time, it's not ISO,   in my current implementation it is:
s := t.Format("2006-01-02 15:04:05.999999999 -0700 MST")

Comment (andreaangiolillo) [reply cfn-resources/serverless-instance/cmd/resource/resource.go]: good point. I updated it to use `util.TimePtrToStringPtr` which set the format to `RFC3339`

**Review by lantoli — APPROVED** [review]

**Pull request merged**